### ISSUE TITLE
performance issues on a single MPI process

### ISSUE THREAD

**Issue by coquelin77** [helmholtz-analytics/heat#789]
**Description**
Simple operations are showing a massive slowdown vs numpy/torch when run on a single mpi process. 

**To Reproduce**
1. Which module/class/function is affected?
    - so far, it seems to only be the binary ops, however it may effect other things as well. 
2. What are the circumstances under which the bug appears?
    - single mpi process runs on CPU show the behavior most clearly 
3. What is the exact error message / erroneous behavior?
    - the times of all of the processes should be very similar

**Expected behavior**
the times of all three packages should be similar. 

**Illustrative**

```python
  import heat as ht
  import numpy as np
  import time
  arrsize = 100000000
  times = []
  for _ in range(10):
      a = ht.ones(arrsize)#, split=0)
      b = ht.ones(arrsize)#, split=0)
      t1 = time.perf_counter()
      c = a * b + a ** 2 + b ** 2 + a - b
      t2 = time.perf_counter()
      times.append(t2 - t1)
  print(np.mean(times))
```
1 process heat: `2.7430931525974303`
1 process torch: `1.3330379242979689`

with a bit more digging i can see that this might be happening within the `a ** 2` operation.

**Version Info**
Current master (184f112)

**Additional comments**
Thank you to @fschlimb for pointing this out

**Comment by ClaudiaComito** [comment]
Torch binary operations seem to be much more efficient on scalars than on torch tensors. However, Heat always casts scalars to DNDarrays, hence internally Heat's binary ops are always between torch tensors.

Using the script above, but `c = a**2`.   On my laptop, with `mpirun -n 1 `, I get the following:
```
heat times:  0.8049912449333334
numpy times:  0.70067716212
torch times:  0.5644471529142856
```

However, if I replace `a**2` with `a**torch.tensor(2)` (or `a**np.array(2)` for numpy) I get the following:

```
heat times:  0.814409041133333
numpy times:  0.7144210714799999
torch times:  0.7410358746
```
Interesting that for numpy it doesn't make a difference.

I'll think about how to simplify the `__binary_op()` checks and handling of scalars, or @coquelin77 do you want to take over?

**Comment by ClaudiaComito** [comment]
Some single-process test runs:

| Operation  |  Heat  runtime (s) |   NumPy  runtime (s)|  PyTorch runtime (s)|   Heat/PyTorch |
| :--  | :-------------: | :----------: | :-----------: | :-----------: |
| `a**2`  | 0.81  | 0.70    |  0.57   | 1.4   |
| `a*a`    |  0.22 |  0.36  |   0.32  |  0.7 |
| `a*2`     | 0.20 |  0.33  |   0.30  |  0.7 |
|  `a+b`  |  0.23 |  0.36  | 0.33    |  0.7 |
|  `a+2`  |  0.20 |  0.31  | 0.38    | 0.5 |
| `a*b + a**2 + b**2 + a-b`  | 2.69 |   2.49 |   2.19  | 1.2 |
|    `a*b + a*a + b*b + a-b` | 1.55 |  1.82  |    1.72 |  0.9 |

We can think about exploiting `torch.pow` better, but I don't see a reason to make big changes in our `__binary_op()` at this time.

**Comment by ClaudiaComito** [comment]
Some single-process test runs (all arrays `float32`):

| Operation  |  Heat  runtime (s) |   NumPy  runtime (s)|  PyTorch runtime (s)|   Heat/PyTorch |
| :--  | :-------------: | :----------: | :-----------: | :-----------: |
| `a**2`  | 0.83  | 0.59    |  0.49   | 1.69   |
| `a*a`    |  0.25 |  0.24  |   0.24  |  1.04 |
| `a*2`     | 0.25 |  0.24  |   0.25  |  1 |
|  `a+b`  |  0.28 |  0.27  | 0.26    |  1.08 |
|  `a+2`  |  0.22 |  0.22  | 0.22    | 1 |
| `a*b + a**2 + b**2 + a-b`  | 2.75 |   2.01 |   1.88  | 1.46 |
|    `a*b + a*a + b*b + a-b` | 1.55 |  1.31  |    1.38 |  1.12 |

**Comment by coquelin77** [comment]
what is your data size for the benchmarks?

**Comment by ClaudiaComito** [comment]
> what is your data size for the benchmarks?

Hey Daniel, I'm running the exact same example above for different operations, only difference I'm setting `dtype`.

**Comment by coquelin77** [comment]
okay i was just curious. the `**` is very bad. but this is a direct call to the binary op. im not sure why it would take so much longer since it should just call torch directly. the only way that i can think of to really increase the speed would be to put a try-except block at the top of the binary op and attempt it without the isinstance checks or adjustments. or another form of early out might work too

**Comment by coquelin77** [comment]
or maybe we just dont use binary op in pow

**Comment by fschlimb** [comment]
Hm, the difference I see is not just a few %, it's a factor of 3 slower than numpy and 10x slower than torch. Something else might be going on.

@ClaudiaComito If you give me your code and tell me which versions of numpy, torch, heat etc you use (or a conda env yaml file) I could generate the same data for one or 2 of our machines and also compare different python envs.

**Comment by coquelin77** [comment]
I cant speak for Claudia but I did a bit of testing and with a minor rewrite of the pow function i was able to close the gap a bit:
### orginal
```
heat 19.422133721411228
np 4.610380478855222
torch 5.885436344053597
```
### new
```
heat 6.7831931237131355
np 4.600037970021367
torch 5.761834943573922
```
these are measurements of the full operation (`c = a * b + a ** 2 + b ** 2 + a - b`) with `arrsize=1_000_000_000` to make the differences more obvious. This is with torch 1.7.0, numpy 1.19.1, and a branch off the current master for heat. I am not sure if this is the Intel numpy.

there is still some space for improvement but there are some checks that we have to perform at the top of these operations to make sure that everything is in correct. 

If anyone wants to play around with the new code, you can find it on this branch: bug/789-pow-binary-op-performance

These optimizations will only take effect if one of the inputs is a DNDarray.

**Comment by ClaudiaComito** [comment]
> Hm, the difference I see is not just a few %, it's a factor of 3 slower than numpy and 10x slower than torch. Something else might be going on.
> 
> @ClaudiaComito If you give me your code and tell me which versions of numpy, torch, heat etc you use (or a conda env yaml file) I could generate the same data for one or 2 of our machines and also compare different python envs.

@fschlimb at last here's my code, which I ran on the current main branch, and my .yaml. Looking forward to your tests. Have a nice weekend!

[binop_test.zip](https://github.com/helmholtz-analytics/heat/files/6641634/binop_test.zip)

**Comment by fschlimb** [comment]
@ClaudiaComito Thanks. For some reason the numbers look reasonable today. Not only the env that uses similar package versions as you do, but only in my old one (but I have updated HeAT). Not sure what happened. Let's see if it comes up again.

Just fyi: the script did not report correct times because the `times` list does not get reset between array implementations.

I also made some basic scalability experiments (2 nodes). By default, there is barely any speedup when adding another node. Something seems to be going wrong with OpenMP. In my case default numthreads should be 4; if I set it to 4 on single-node the performance is indeed similar to not setting OMP_NUM_THREADS. It scales well with OMP_NUM_THREADS set (to 1,2 and 4) on each node but not without. I see this when running MPI on one as well as on two nodes. OpenMP scalability seems fine on a single node as well. Any idea?

**Comment by coquelin77** [comment]
The OpenMP issue is interesting. We do not do anything with threading and rely on torch to handle this for us. During some internal testing, we found that the optimum configuration for our testing cluster was 2 MPI processes each with 12 OMP threads, however this is likely unique to our system. 

Using this setup, i have not been able to reproduce the speedup issues when adding nodes. If the split axis is set in the example above, I found that (with and without the efficiency fixes done in #793) the calculation has a speedup proportional to the number of MPI processes. if `split=None` then the calculation is run on all processes concurrently and there will be no intentional speedup.

**Comment by ClaudiaComito** [comment]
Performance improvement in (upcoming) version 1.3:

![pow](https://user-images.githubusercontent.com/39374113/234465310-1a0eb775-8222-418e-9efb-1c6734610990.png)
![pow_zoomin](https://user-images.githubusercontent.com/39374113/234465324-9dadc89f-df55-46a5-986b-229d175ef645.png)
